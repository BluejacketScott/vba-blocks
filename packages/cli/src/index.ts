export { default as env } from './env';
export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest,
  loadManifest
} from './manifest';
export { Config, loadConfig } from './config';
export { default as resolve } from './resolve';
