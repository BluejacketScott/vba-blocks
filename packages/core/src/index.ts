export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest
} from './manifest';
export { Config } from './config';
