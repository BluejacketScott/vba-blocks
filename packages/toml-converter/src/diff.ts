import { diff as deepDiff, Index, Path } from 'deep-diff';
import traverse from 'traverse';
import clone from 'clone';

/*
  deep-diff has an admittedly simple approach to handling array changes,
  this adds a system for better determining array moves
  
  1. Walk lhs object and identify arrays
  2. Find removed, added, and moved values
  3. Store change information and apply move/remove operations
  4. Diff modified object to identify remaining changes

  Adds a new kind of change: 'R' (for reorder)

  ```js
  {
    kind: 'R',
    path: ['path', 'to', 'array'],
    indices: [2,1,0]
  }
  ```

  Above = Reorder from [0,1,2] to [2,1,0]
*/

export interface Change {
  kind: 'N' | 'D' | 'E' | 'A' | 'R';
  path?: Path;
  lhs?: any;
  rhs?: any;
  index?: number;
  indices?: number[];
  item?: Change;
}

// There are probably heuristics to identify both moves in the array and object changes,
// but for simplicity an id function is provided to identify values in the array
export type getObjectId = (value: any, path: Path, found_in: any[]) => any;

export default function diff(
  lhs: any,
  rhs: any,
  options: { getId?: getObjectId } = {}
): Change[] {
  const array_changes = diffArray(lhs, rhs, options);
  lhs = applyArrayChanges(lhs, array_changes);

  const deep_changes = deepDiff(lhs, rhs) || [];
  return [...array_changes, ...deep_changes];
}

/*
  [1,2,3,4,5] to [6,4,3,2,6,7]

  1. Remove - [2,3,4] (at index)
  2. Reorder - [4,3,2] (by index)
  3. Add - [4,3,2,6,7] (at index)
*/
function diffArray(
  lhs: any,
  rhs: any,
  options: { getId?: getObjectId }
): Change[] {
  const { getId = (value: any) => value } = options;

  const paths: Path[] = traverse(lhs).paths();
  let changes: Change[] = [];

  for (const path of paths) {
    const previous = get(lhs, path);
    const next = get(rhs, path);

    if (!next || !Array.isArray(previous)) continue;

    const previous_ids = previous.map((value: any) =>
      getId(value, path, previous)
    );
    const next_ids = next.map((value: any) => getId(value, path, next));

    const remove = indexedDifference(previous_ids, next_ids);
    const add = indexedDifference(next_ids, previous_ids);
    const reorder = reordered(
      removeByIndex(previous_ids, remove),
      removeByIndex(next_ids, add)
    );

    remove.forEach(index => changes.push({ kind: 'D', path, index }));
    if (reorder.length) changes.push({ kind: 'R', path, indices: reorder });
    add.forEach(index =>
      changes.push({
        kind: 'A',
        path,
        index,
        item: { kind: 'N', rhs: next[index] }
      })
    );
  }

  return changes;
}

function applyArrayChanges(value: any, changes: Change[]): any {
  value = clone(value);

  for (const change of changes) {
    const { kind, path, index, indices, item } = change;
    const array = get(value, change.path!);
    if (kind === 'D') {
      array.splice(index, 1);
    } else if (kind === 'A') {
      array.splice(index, 0, item!.rhs);
    } else if (kind === 'R') {
      const original = array.slice();
      indices!.forEach((to, from) => (array[to] = original[from]));
    }
  }

  return value;
}

function difference(values: any[], reference: any[]): any[] {
  return values.filter((value: any) => !reference.includes(value));
}

function indexedDifference(values: any[], reference: any[]): number[] {
  return values.reduce((memo: number[], value: any, index: number) => {
    if (!reference.includes(value)) memo.push(index);
    return memo;
  }, []);
}

function removeByIndex(values: any[], to_remove: number[]): any[] {
  return values.filter(
    (value: any, index: number) => !to_remove.includes(index)
  );
}

function reordered(from: any[], to: any[]): number[] {
  const found: boolean[] = [];
  const reorder = from.map((value: any) => {
    let index = to.indexOf(value);
    while (index >= 0 && found[index]) {
      index = to.indexOf(value, index + 1);
    }

    if (index < 0)
      throw new Error(`Could not find moved value "${value}" in destination`);
    found[index] = true;

    return index;
  });

  const noop = reorder.every((to, from) => to === from);

  return noop ? [] : reorder;
}

function get(object: any, path: Path): any {
  for (const part of path) {
    if (!object) return undefined;
    object = object[part];
  }
  return object;
}
