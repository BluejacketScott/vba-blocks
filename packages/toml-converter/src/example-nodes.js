// TEMP

[
  {
    type: 'ObjectPath',
    value: ['package'],
    line: 1,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: 'package-name',
      line: 2,
      column: 8
    },
    line: 2,
    column: 1,
    key: 'name'
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: '0.0.0',
      line: 3,
      column: 11
    },
    line: 3,
    column: 1,
    key: 'version'
  },
  {
    type: 'Assign',
    value: {
      type: 'Array',
      value: [
        {
          type: 'String',
          value: 'Author 1',
          line: 4,
          column: 12
        }
      ],
      line: 4,
      column: 11
    },
    line: 4,
    column: 1,
    key: 'authors'
  },
  {
    type: 'ObjectPath',
    value: ['dependencies'],
    line: 12,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: '1.0.0',
      line: 13,
      column: 5
    },
    line: 13,
    column: 1,
    key: 'a'
  },
  {
    type: 'Assign',
    value: {
      type: 'InlineTable',
      value: [
        {
          type: 'InlineTableValue',
          value: {
            type: 'String',
            value: '...git',
            line: 14,
            column: 13
          },
          line: 14,
          column: 7,
          key: 'git'
        },
        {
          type: 'InlineTableValue',
          value: {
            type: 'String',
            value: 'weird',
            line: 14,
            column: 38
          },
          line: 14,
          column: 25,
          key: 'spacing'
        }
      ],
      line: 14,
      column: 5
    },
    line: 14,
    column: 1,
    key: 'b'
  },
  {
    type: 'ObjectPath',
    value: ['dependencies', 'c'],
    line: 16,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'String',
      value: '...path',
      line: 17,
      column: 8
    },
    line: 17,
    column: 1,
    key: 'path'
  },
  {
    type: 'ArrayPath',
    value: ['array'],
    line: 19,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'Integer',
      value: 0,
      line: 20,
      column: 9
    },
    line: 20,
    column: 1,
    key: 'index'
  },
  {
    type: 'ArrayPath',
    value: ['array'],
    line: 22,
    column: 1
  },
  {
    type: 'Assign',
    value: {
      type: 'Integer',
      value: 1,
      line: 23,
      column: 9
    },
    line: 23,
    column: 1,
    key: 'index'
  }
];
