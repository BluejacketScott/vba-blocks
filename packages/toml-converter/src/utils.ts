export function toLines(value: string): string[] {
  return value.replace(/\r/g, '').split('\n');
}
