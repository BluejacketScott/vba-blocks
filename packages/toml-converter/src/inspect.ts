import { Path, Index } from 'deep-diff';
import { graceful as detectNewline } from 'detect-newline';
import { Change } from './diff';
import { toLines } from './utils';

// Use __internal__ parsing from toml library
// (this should not be assumed to be stable and a specific version may need to be pinned)
export interface ParsedNode {
  type:
    | 'ObjectPath'
    | 'ArrayPath'
    | 'Assign'
    | 'InlineTable'
    | 'InlineTableValue'
    | 'Array'
    | 'String'
    | 'Float'
    | 'Integer'
    | 'Boolean'
    | 'Date';

  line: number;
  column: number;
  value: any;
  key?: string;
}

const parseToml: (toml: string) => ParsedNode[] = require('toml/lib/parser')
  .parse;

/*
  Approach: Arrange nodes into AST tree and add end information

  Example:
  ```toml
    123456789
  1 [package]
  2 name = "value"
  3 authors = ["a", "b"]
  4
  5 [dependencies.a]
  6 git = { path = "...", tag = "v1.0.0" }
  ```

  ```js
  {
    package: {
      key: {
        start: { line: 1, column: 1 },
        end: { line: 1, column 9 },
        parsed: (from toml = package)
      },
      value: {
        start: { line: 2, column: 1 },
        end: { line: 4, column: 1 },
        parsed: (from toml)
      },
      children: {
        name: {
          key: { start: { ... }, end: { ... }, parsed: (name)},
          value: { start: { ... }, end: { ... }, parsed: (value)}
        },
        authors: {
          key: { ... },
          value: { ... },
          children: [
            { key: null, value: { ..., parsed: (a) } },
            { key: null, value: { ..., parsed: (b) } }
          ]
        }
      }
    },
    dependencies: {
      key: null,
      value: { start: { line: 5 }, ...},
      children: {
        a: {
          key: { ... },
          value: { ... },
          children: {
            git: {
              key: { end: { line: 6, column: 3 } },
              value: { ... },
              children: {
                path: {
                  key: { ... },
                  value: { ..., parsed: (...) }
                },
                tag: {
                  key: { ... },
                  value: { ..., parsed: (v1.0.0) }
                }
              }
            }
          }
        }
      }
    }
  }
  ```
*/

export type AST = KeyedNodes;

export interface KeyedNodes {
  [key: string]: Node;
}
export type IndexedNodes = Node[];

export interface Node {
  key?: Value;
  value: Value;
  children?: KeyedNodes | IndexedNodes;
}

export interface Value {
  start: Position;
  end: Position;
  parsed?: ParsedNode;
}

export interface Position {
  line: number;
  column: number;
}

export default function inspect(toml: string): AST {
  const lines = toLines(toml);
  const nodes: ParsedNode[] = parseToml(lines.join('\n'));
  return <KeyedNodes>walk(nodes, lines);
}

function walk(parsed: ParsedNode[], lines: string[]): KeyedNodes {
  let nodes: KeyedNodes = {};
  let current: KeyedNodes = nodes;

  for (const [index, node] of parsed.entries()) {
    let { line, column } = node;
    let key: null | Value;
    let value: Value;
    let children: KeyedNodes | IndexedNodes;
    let result: Node;

    let next_node: ParsedNode | undefined;
    let end_line: number;
    let end_of_key: number;
    let end_of_value: number;

    switch (node.type) {
      case 'ObjectPath':
      case 'ArrayPath':
        // Parsed ArrayPath points to first line of values (not key)
        // if (node.type === 'ArrayPath') line -= 1;

        // { "value": ["package"] }
        next_node = parsed.find(
          (node, _index) => _index > index && isPathNode(node)
        );
        end_line = next_node ? next_node.line - 1 : lines.length;

        // TODO something is wrong with end.column position (-> line length)

        result = {
          key: {
            start: { line, column },
            end: { line, column: lines[line - 1].length },
            parsed: node
          },
          value: {
            start: { line: line + 1, column: 0 },
            end: { line: end_line, column: lines[end_line - 1].length }
          },
          children: {}
        };

        current = <KeyedNodes>result.children;

        node.type === 'ObjectPath'
          ? set(nodes, node.value, result)
          : push(nodes, node.value, result);

        break;

      case 'Assign':
        // { "key": "name", "value": { "type": "String", "value": "package-name" }
        next_node = parsed[index + 1];
        end_line = next_node ? next_node.line - 1 : lines.length;
        end_of_key = findCharColumn(lines[line - 1], '=', {
          from: column,
          fallback: column + node.key!.length
        });

        result = {
          key: {
            start: { line, column },
            end: { line, column: end_of_key },
            parsed: node
          },
          value: {
            start: { line, column: end_of_key + 1 },
            end: { line: end_line, column: lines[end_line - 1].length },
            parsed: node.value
          },
          children: toChildren(node.value, lines)
        };

        set(current, [node.key!], result);

        break;

      case 'InlineTableValue':
        // { "key": "name", "value": { "type": "String", "value": "package-name" }
        end_of_key = findCharColumn(lines[line - 1], '=', {
          from: column,
          fallback: column + node.key!.length
        });
        end_of_value = findCharColumn(lines[line - 1], [',', '}'], {
          from: end_of_key
        });

        result = {
          key: {
            start: { line, column },
            end: { line, column: end_of_key },
            parsed: node
          },
          value: {
            start: { line, column: end_of_key + 1 },
            end: { line, column: end_of_value },
            parsed: node.value
          },
          children: toChildren(node.value, lines)
        };

        set(current, [node.key!], result);

        break;

      default:
        throw new Error(
          `Unrecognized/unhandled parsed node with type "${node.type}"`
        );
    }
  }

  return nodes;
}

function toChildren(
  value: ParsedNode,
  lines: string[]
): KeyedNodes | IndexedNodes | undefined {
  switch (value.type) {
    case 'InlineTable':
      return walk(value.value, lines);
    case 'Array':
      return <IndexedNodes>value.value.map((child: ParsedNode) => {
        const { line, column } = child;
        const children = toChildren(child, lines);

        const end_line = Array.isArray(children)
          ? children[children.length - 1].value.end.line
          : line;
        const search_column = Array.isArray(children)
          ? children[children.length - 1].value.end.column
          : column;
        const end_of_value = findCharColumn(lines[end_line - 1], ',', {
          from: search_column
        });

        const value: Value = {
          start: { line, column },
          end: { line, column: end_of_value },
          parsed: child
        };

        return { value, children };
      });
    default:
      return;
  }
}

export function get(ast: AST, path: Path): Node | null {
  if (!path.length) return null;

  const [root, ...nested] = path;
  let node: Node = ast[root];

  for (const part of nested) {
    const children = node && node.children;
    if (!children) return null;

    node = Array.isArray(children)
      ? children[<number>part]
      : children[<string>part];
  }

  return node;
}

export function has(ast: AST, path: Path): boolean {
  return !!get(ast, path);
}

function set(ast: AST, path: Path, node: Node): Node {
  if (path.length === 1) {
    return (ast[path[0]] = node);
  }

  const key: string = <string>path[path.length - 1];
  const parent_path = path.slice(0, -1);
  let parent = get(ast, parent_path);

  if (!parent) {
    parent = {
      value: {
        start: node.key!.start,
        end: node.value.end
      },
      children: {}
    };

    set(ast, parent_path, parent);
  }
  if (!parent.children) {
    parent.children = {};
  }

  // Each time a new value is add to parent,
  // Ensure end location is up-to-date
  if (node.value.end.line > parent.value.end.line) {
    parent.value.end = node.value.end;
  }

  return ((<KeyedNodes>parent.children)[key] = node);
}

function push(ast: AST, path: Path, node: Node): Node {
  let parent = get(ast, path);

  if (!parent) {
    parent = {
      value: {
        start: node.key!.start,
        end: node.value.end
      },
      children: []
    };

    set(ast, path, parent);
  }
  if (!parent.children) {
    parent.children = [];
  }

  // Each time a new value is pushed onto path,
  // update the end position of the parent's value
  parent.value.end = node.value.end;
  (<IndexedNodes>parent.children!).push(node);

  return node;
}

function isPathNode(node: ParsedNode): boolean {
  return node.type === 'ObjectPath' || node.type === 'ArrayPath';
}

const QUOTE = '"';
const ESCAPE = '\\';

function findCharColumn(
  value: string,
  search: string | string[],
  options?: { from?: number; fallback?: number }
): number {
  const { from = 1, fallback = value.length } = options || {};

  if (!Array.isArray(search)) search = [search];
  for (const check of search) {
    let quoted = false;
    for (let i = from; i <= value.length; i++) {
      const char = value.charAt(i - 1);
      if (char === check && !quoted) return i;

      if (!quoted && char === QUOTE) quoted = true;
      if (quoted && char === QUOTE && value.charAt(i - 2) !== ESCAPE)
        quoted = false;
    }
  }

  return fallback;
}
