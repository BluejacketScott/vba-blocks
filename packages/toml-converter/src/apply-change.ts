import { AST, Value, get } from './inspect';
import { Change } from './diff';

export default function applyChange(
  toml: string[],
  ast: AST,
  change: Change
): string[] {
  // TODO
  switch (change.kind) {
    case 'N':
      break;
    case 'D':
      break;
    case 'E':
      const node = get(ast, change.path!)!;
      replace(toml, node.value, change.rhs);

      break;
    case 'A':
      break;
    case 'R':
      break;
  }

  // TODO
  return toml;
}

function isMultiline(part: Value): boolean {
  return part.end.line > part.start.line;
}

function replace(toml: string[], value: Value, replacement: any) {
  if (isMultiline(value) || isObject(replacement)) return;

  const index = value.start.line - 1;
  const line = toml[index];
  toml[index] =
    line.substr(0, value.start.column - 1) +
    replacement +
    line.substr(value.end.column);
}

function toInlineTable(value: any): string {
  // TODO
  return '';
}

function isObject(value: any): boolean {
  return value != null && typeof value === 'object';
}
