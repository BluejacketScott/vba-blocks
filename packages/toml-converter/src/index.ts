import { parse as parseToml } from 'toml';
import { diff } from 'deep-diff';
import inspect from './inspect';

export function parse(toml: string): any {
  return parseToml(toml);
}

export function convert(value: any): string {
  // TODO
  return '';
}

export function patch(toml: string, value: any): string {
  const existing = parse(toml);
  const changes = diff(existing, value);

  const ast = inspect(toml);
  ast.applyChanges(changes);

  return ast.toToml();
}
