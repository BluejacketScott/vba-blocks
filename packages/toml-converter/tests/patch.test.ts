import dedent from 'dedent';
import { parse, patch } from '../src';

export const fixture = dedent`
  [package]
  name = "package-name" # comment
  version = "0.0.0"
  authors = ["Author 1"]

  # 6
  # 7
  # 8

  # 10

  [dependencies] # 12
  a = "1.0.0"
  b = { git = "...git"  , spacing =    "weird"}#comment

  [dependencies.c] #16
  path = "...path"

  [references.nested] #19
  guid = "{...}"

  [[array]] # 22
  index = 0

  [[array]]
  index = 1
  
  [[array]]
  index = 2`;

test('patch', () => {
  const value = parse(fixture);
  value.package.authors.push('Author 2');
  value.package.custom = true;
  value.dependencies.a = { version: '2.0.0' };
  value.dependencies.b.optional = true;
  value.dependencies.c.path = '...newpath';
  value.dependencies.d = '0.0.0';
  value.dependencies.e = { version: '1.0.0', optional: true };

  value.array.pop();
  value.array[0].other = 'yes';
  value.array.unshift({ index: -1 });

  const result = patch(fixture, value, {
    getId: value => (typeof value !== 'object' ? value : value.index)
  });

  expect(result).toMatchSnapshot();
});
