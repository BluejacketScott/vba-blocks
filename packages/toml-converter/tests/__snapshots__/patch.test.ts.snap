// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patch 1`] = `
"[package]
name = \\"package-name\\" # comment
version = \\"0.0.0\\"
authors = [\\"Author 1\\"]

# 6
# 7
# 8

# 10

[dependencies] # 12
a = \\"1.0.0\\"
b = { git = \\"...git\\"  , spacing =    \\"weird\\"}#comment

[dependencies.c] #16
path = \\"...path\\"

[references.nested] #19
guid = \\"{...}\\"

[[array]] # 22
index = 0

[[array]]
index = 1

[[array]]
index = 2"
`;
